Add module on odo storage.

boczkowska opened the pull request.

dharmit commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,30 @@
+
+When ephemeral storage is enabled, the cluster stores temporary information that does not need to persist between the restarts.
+
+Persistent storage is enabled by default. 
```
This changed with https://github.com/openshift/odo/pull/4360. Ephemeral is enabled by default. The user has to set it to false to use persistent storage.
```sh
$ odo preference -h | grep Ephemeral
 Ephemeral - If true odo will create a emptyDir volume to store source code (Default: true)
```

dharmit commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,30 @@
+$ odo preference set ephemeral true
+----
+
+. Push the changes to the cluster:
```
We don't necessarily need `odo push` to apply Ephemeral vs. Persistent thing. However, it is true that only `odo push` would help the user see the change take place on their component.

Thing is, this being a Preference setting, it applies to all components currently deployed by the user as well as ones they would deploy in future.

I'm more inclined towards using a statement like below:
> Now doing `odo push` for any of your component(s) would result in odo creating an Ephemeral storage for it.

Finally, since this section of doc (`To set the ephemeral storage:`) is written to help user switch from Persistent to Ephemeral, can we modify it to help them convert from Ephemeral to Persistent instead? Asking for this becasue Ephemeral is default.

boczkowska replied:
Thank you, @dharmit 
>Thing is, this being a Preference setting, it applies to all components currently deployed by the user as well as ones they would deploy in future.
It's project-specific, or am I wrong? Each project has its own config.

I've made some changes - added commands on how to bot set and unset, PTAL when you can. 

Although now I'm thinking - maybe it's better to document:
>odo preference view
>odo preference unset ephemeral
>odo push
>oc get po,pvc 
>odo preference set ephemeral
>odo push
>oc get po,pvc 

It's a lot of steps but it will show how to set and check the config and actually see whether the persistent storage has been created.

dharmit commented on modules/developer-cli-odo-basic-odo-cli-commands.adoc:
```diff
@@ -360,6 +360,9 @@ Modify `odo` specific configuration settings within the global preference file.
   odo preference set experimental false
+  
+  # Use ephemeral volumes in the cluster
+  odo preference set ephemeral true
```
Shouldn't this example about setting the value to `false` since the default is `true`? Or am I looking it the wrong way? :thinking:

dharmit commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,61 @@
+
+When persistent storage is enabled, the cluster stores the information between the restarts. 
+
+When ephemeral storage is enabled, the cluster stores temporary information that does not need to persist between the restarts.
```
Between this line and the one above, we have classified "information" vs. "temporary information". It's more like, odo treats stuff as temporary when using ephemeral and as not temporary when not using ephemeral. IMO, we could simply modify the statement for ephemeral storage to below:
`When ephemeral storage is enabled, the cluster does not store/persist the information between the restarts.`

dharmit commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,61 @@
++
+[source,terminal]
+----
+$ odo preference unset ephemeral
```
Simply unsetting the value will still use Ephemeral storage because it considers default value for Ephemeral when none is provided. And default value is `true`.

To ensure that odo uses Persistent storage for components, user needs to set the value to false:
`odo preference set Ephemeral false`.

This comment might help simplify things: https://github.com/openshift/odo/pull/4360#issuecomment-762162203.

dharmit reviewed (changes requested):
Sorry about the confusion with `odo preference set` thingy. Hope the reference comment I have pasted helps ease things out. :smile:

dharmit commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,68 @@
++
+[source,terminal]
+----
+$ odo preference set ephemeral
```
Running `odo preference set ephemeral` after doing `odo preference set Ephemeral false`, as mentioned in a step earlier, is going to throw an error. It's going to fail with `✗  please provide a parameter name and value`.

To move from using Persistent storage (which can be enabled *only* by doing `odo preference set Ephemeral false`) to using Ephemeral storage, one has to do either `odo preference unset Ephemeral` (because when you unset its value, odo assumes its default value which is `true`), or `odo preference set Ephemeral true` (so that you don't leave any room for confusion about what odo's going to read as far as Ephemeral preference's setting is concerned.)

dharmit reviewed (approved):
Looks good. Thanks for bearing with me !

ahardin-rh commented on modules/developer-cli-odo-switching-between-ephemeral-and-persistent-storage.adoc:
```diff
@@ -0,0 +1,61 @@
++
+[source,terminal]
+----
+odo preference view
```
I think this needs a command prompt $ before the command itself. Same for the other commands in this module.

boczkowska replied:
@ahardin-rh thanks! Added the prompt.

ahardin-rh commented:
@boczkowska Just one comment from me. Otherwise, looks great!

boczkowska commented:
> @boczkowska Just one comment from me. Otherwise, looks great!

@ahardin-rh Thank you for the review, I've added the prompt and squashed the commits. Could I ask you to merge and cherry-pick?

ahardin-rh commented:
/cherrypick enterprise-4.7

ahardin-rh commented:
/cherrypick enterprise-4.8
